### ISSUE TITLE
PSP + injection interaction issues

### ISSUE THREAD

**Issue by grampelberg** [linkerd/linkerd2#2886]
### What problem are you trying to solve?

PSP does more than validate pod specs, it also mutates them. Because auto-inject adds a pod and comes afterwards, specific PSP policies can cause pods to be unable to be scheduled. While there is an upcoming feature (1.15?) in k8s to make this better, it will be painful for Linkerd users that also have PSP enabled for awhile.

### How should the problem be solved?

..... unclear

**Comment by grampelberg** [comment]
@alpeb could you explain a little more about what is happening here? It feels like a good use for `check` or `check --proxy` but I'm not sure what it should check.

**Comment by alpeb** [comment]
@grampelberg Your initial post is spot-on. #2767 has more details about this.

To recap, during the PSP mutation phase, if a container doesn't have some condition specified in the PSP, the container is mutated by including that condition, e.g. `readOnlyRootFilesystem: true`.

As you pointed out, that mutating phase won't run on the auto-injected sidecars. To continue with the example, `readOnlyRootFilesystem` will have its default value `false`, instead of the one that should have been set during the mutation phase. This will make the PSP validation to fail.

The current workaround for k8s < 1.15 is to use `inject --manual` when PSP is enabled.

This does sound like a great use case for `check`; If PSP is enabled and k8s version < 1.15 then show a warning saying auto-injection can't work, and `inject --manual` should be used instead.

**Comment by ihcsim** [comment]
If possible, I think we should support psp + auto-inject for pre-1.15 users. Otherwise, auto-inject will be useless in psp-enabled clusters. Not 100% sure, but I *think* the uninjected workload YAML included in the admission request to the proxy injector will include the `kubernetes.io/psp: <selected-policy>` annotation, allowing us to validate if the proxy and proxy-init comply with the rules in the selected policy (e.g. `NET_ADMIN` capability for proxy-init), and then either injecting some defaults or the selected policy rules into the proxy container's security context.

**Comment by alpeb** [comment]
That's a good call @ihcsim. I confirmed the annotation is indeed being received by the MWC.

So we'd need to augment the proxy injector's RBAC to allow read access to the PSPs, and then basically mimic (or directly leverage if possible) the PSP mutating functionality from
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/security/podsecuritypolicy/admission.go

@grampelberg WDYT?

**Comment by grampelberg** [comment]
How long do y'all think it'd take?

**Comment by ihcsim** [comment]
> and then basically mimic (or directly leverage if possible) the PSP mutating functionality 

I don't think we should try to copy/mimic what the psp admission controller is doing, as it looks complicated and who knows if there are bugs in there. Is it possible for us to identify some default security context for the proxy and proxy-init containers? E.g., @alpeb per your previous comment, we can always set `allowPrivilegeEscalation: false` and `readOnlyRootFileSystem: true`, which I think is a good practice anyway.

I tested the control plane with this security context couple weeks ago and they seem to work fine:
```
allowPrivilegeEscalation: false
readOnlyRootFilesystem: true
allowedCapabilities: # for proxy-init
- NET_ADMIN
```

The Linkerd CNI seems slightly different. See https://github.com/linkerd/linkerd2/issues/2819#issuecomment-491984236.

> How long do y'all think it'd take?

Depends on if we want to dig into the psp admission controller code or just define defaults. Most of the effort will be on testing (on a cloud provider. Last I checked, the psp on Minikube seems flaky.) Hopefully, a couple days.

**Comment by grampelberg** [comment]
As mutation is the real problem for us, if we can get the common mutations taken care of, validate the policy and fail injecting if the policy wouldn't allow it .... that feels like a pretty good solution for now?

**Comment by alpeb** [comment]
Ok adding the most restrictive conditions that our sidecar allows should allow us to pass the most commons cases indeed. I'll open a PR.

**Comment by ihcsim** [comment]
@alpeb @grampelberg Let me know if we want to keep this issue open as a way to track the k8s 1.5 fix, or if we can close it with #2911 merged.

**Comment by alpeb** [comment]
It's ok for me to close it.

**Comment by ihcsim** [comment]
Fixed by #2911.
